File: selectionsort.py



def selection_sort(inp):
    for i in range(len(inp)-1, -1, -1):
        maximum = maxEl(inp[:i+1])
        temp = inp[maximum]
        inp[maximum] = inp[i]
        inp[i] = temp

    return inp


def maxEl(inp):
    maxI = 0
    for i in range(0, len(inp)):
        if inp[i] >= inp[maxI]:
            maxI = i
    
    return maxI



def main():
    print(selection_sort([3,2,1,6,5]))

if __name__=="__main__": 
    main() 

File: selectionsort.hs

-- todo: integrated maximum and indexof function to make it more efficient

selection_sort :: [Int] -> [Int]
selection_sort arr
    | null arr = []
    | otherwise = (selection_sort (start ++ end)) ++ [maxi]
    where
        maxi = maximum arr
        index = getindexof maxi arr
        start = take index arr
        end = drop (index+1) arr

getindexof :: Int -> [Int] -> Int
getindexof target (x:xs)
    | x == target = 0
    | otherwise = (getindexof target xs) + 1

